Register format ``html`` to ``Cosmology.write`` with nice mathjax

Cosmology can now read and write to files.
It would be nice to register with ``Cosmology.write`` a  method for exporting a Cosmology to a HTML table.
There are good examples of IO with Cosmology at https://github.com/astropy/astropy/tree/main/astropy/cosmology/io
and documentation at https://docs.astropy.org/en/latest/cosmology/io.html#cosmology-io

I'm thinking the ``write_html(...)`` method would call ``cosmology.io.table.to_table()``, format the table to nice MathJax or something and then call the `QTable.write(..., format='html')`.

Edit: also, the mathjax representation of each parameter can be stored on the corresponding Parameter object, like how units have the ``format`` argument in [def_unit](https://docs.astropy.org/en/stable/api/astropy.units.def_unit.html#astropy.units.def_unit).

I am writing tests now and it looks like writing fails with the following errors. I am not quite sure why these errors are appearing. 

I based the first test on cosmology\io\tests\test_ecsv.py. 

```
fp = tmp_path / "test_to_html_table_bad_index.html"
write(file=fp)
```


error: 
```
self = <astropy.cosmology.io.tests.test_html.TestReadWriteHTML object at 0x00000175CE162F70>, read = <function ReadWriteDirectTestBase.read.<locals>.use_read at 0x00000175CE2F3280>
write = <function ReadWriteDirectTestBase.write.<locals>.use_write at 0x00000175CE4B9A60>, tmp_path = WindowsPath('C:/Users/jeffr/AppData/Local/Temp/pytest-of-jeffr/pytest-34/test_to_html_table_bad_index_c7')

    def test_to_html_table_bad_index(self, read, write, tmp_path):
        """Test if argument ``index`` is incorrect"""
        fp = tmp_path / "test_to_html_table_bad_index.html"

>       write(file=fp, format="ascii.html")

astropy\cosmology\io\tests\test_html.py:30:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
astropy\cosmology\io\tests\base.py:196: in use_write
    return self.functions["write"](cosmo, *args, **kwargs)
astropy\cosmology\io\html.py:86: in write_table
    table.write(file, overwrite=overwrite, **kwargs)
astropy\table\connect.py:129: in __call__
    self.registry.write(instance, *args, **kwargs)
astropy\io\registry\core.py:354: in write
    return writer(data, *args, **kwargs)
astropy\io\ascii\connect.py:26: in io_write
    return write(table, filename, **kwargs)
astropy\io\ascii\ui.py:840: in write
    lines = writer.write(table)
astropy\io\ascii\html.py:431: in write
    new_col = Column([el[i] for el in col])
astropy\table\column.py:1076: in __new__
    self = super().__new__(
astropy\table\column.py:434: in __new__
    self_data = np.array(data, dtype=dtype, copy=copy)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <Quantity 0. eV>

    def __float__(self):
        try:
            return float(self.to_value(dimensionless_unscaled))
        except (UnitsError, TypeError):
>           raise TypeError('only dimensionless scalar quantities can be '
                            'converted to Python scalars')
E           TypeError: only dimensionless scalar quantities can be converted to Python scalars

astropy\units\quantity.py:1250: TypeError
```

--- a/astropy/cosmology/io/__init__.py
+++ b/astropy/cosmology/io/__init__.py
@@ -5,4 +5,4 @@ Read/Write/Interchange methods for `astropy.cosmology`. **NOT public API**.
 """
 
 # Import to register with the I/O machinery
-from . import cosmology, ecsv, mapping, model, row, table, yaml
+from . import cosmology, ecsv, html, mapping, model, row, table, yaml  # noqa: F401

--- a/astropy/cosmology/io/html.py
+++ b/astropy/cosmology/io/html.py
@@ -1,0 +1,189 @@
+import astropy.cosmology.units as cu
+import astropy.units as u
+from astropy.cosmology.connect import readwrite_registry
+from astropy.cosmology.core import Cosmology
+from astropy.cosmology.parameter import Parameter
+from astropy.table import QTable
+
+from .table import from_table, to_table
+
+# Format look-up for conversion, {original_name: new_name}
+# TODO! move this information into the Parameters themselves
+_FORMAT_TABLE = {
+    "H0": "$$H_{0}$$",
+    "Om0": "$$\\Omega_{m,0}$$",
+    "Ode0": "$$\\Omega_{\\Lambda,0}$$",
+    "Tcmb0": "$$T_{0}$$",
+    "Neff": "$$N_{eff}$$",
+    "m_nu": "$$m_{nu}$$",
+    "Ob0": "$$\\Omega_{b,0}$$",
+    "w0": "$$w_{0}$$",
+    "wa": "$$w_{a}$$",
+    "wz": "$$w_{z}$$",
+    "wp": "$$w_{p}$$",
+    "zp": "$$z_{p}$$",
+}
+
+
+def read_html_table(filename, index=None, *, move_to_meta=False, cosmology=None, latex_names=True, **kwargs):
+    """Read a |Cosmology| from an HTML file.
+
+    Parameters
+    ----------
+    filename : path-like or file-like
+        From where to read the Cosmology.
+    index : int or str or None, optional
+        Needed to select the row in tables with multiple rows. ``index`` can be
+        an integer for the row number or, if the table is indexed by a column,
+        the value of that column. If the table is not indexed and ``index`` is a
+        string, the "name" column is used as the indexing column.
+
+    move_to_meta : bool, optional keyword-only
+        Whether to move keyword arguments that are not in the Cosmology class'
+        signature to the Cosmology's metadata. This will only be applied if the
+        Cosmology does NOT have a keyword-only argument (e.g. ``**kwargs``).
+        Arguments moved to the metadata will be merged with existing metadata,
+        preferring specified metadata in the case of a merge conflict (e.g. for
+        ``Cosmology(meta={'key':10}, key=42)``, the ``Cosmology.meta`` will be
+        ``{'key': 10}``).
+    cosmology : str or |Cosmology| class or None, optional keyword-only
+        The cosmology class (or string name thereof) to use when constructing
+        the cosmology instance. The class also provides default parameter
+        values, filling in any non-mandatory arguments missing in 'table'.
+    latex_names : bool, optional keyword-only
+        Whether the |Table| (might) have latex column names for the parameters
+        that need to be mapped to the correct parameter name -- e.g. $$H_{0}$$
+        to 'H0'. This is `True` by default, but can be turned off (set to
+        `False`) if there is a known name conflict (e.g. both an 'H0' and
+        '$$H_{0}$$' column) as this will raise an error. In this case, the
+        correct name ('H0') is preferred.
+    **kwargs : Any
+        Passed to :attr:`astropy.table.QTable.read`. ``format`` is set to
+        'ascii.html', regardless of input.
+
+    Returns
+    -------
+    |Cosmology| subclass instance
+
+    Raises
+    ------
+    ValueError
+        If the keyword argument 'format' is given and is not "ascii.html".
+    """
+    # Check that the format is 'ascii.html' (or not specified)
+    format = kwargs.pop("format", "ascii.html")
+    if format != "ascii.html":
+        raise ValueError(f"format must be 'ascii.html', not {format}")
+
+    # Reading is handled by `QTable`.
+    with u.add_enabled_units(cu):  # (cosmology units not turned on by default)
+        table = QTable.read(filename, format="ascii.html", **kwargs)
+
+    # Need to map the table's column names to Cosmology inputs (parameter
+    # names).
+    # TODO! move the `latex_names` into `from_table`
+    if latex_names:
+        table_columns = set(table.colnames)
+        for name, latex in _FORMAT_TABLE.items():
+            if latex in table_columns:
+                table.rename_column(latex, name)
+
+    # Build the cosmology from table, using the private backend.
+    return from_table(table, index=index, move_to_meta=move_to_meta, cosmology=cosmology)
+
+
+def write_html_table(cosmology, file, *, overwrite=False, cls=QTable, latex_names=False, **kwargs):
+    r"""Serialize the |Cosmology| into a HTML table.
+
+    Parameters
+    ----------
+    cosmology : |Cosmology| subclass instance file : path-like or file-like
+        Location to save the serialized cosmology.
+    file : path-like or file-like
+        Where to write the html table.
+
+    overwrite : bool, optional keyword-only
+        Whether to overwrite the file, if it exists.
+    cls : |Table| class, optional keyword-only
+        Astropy |Table| (sub)class to use when writing. Default is |QTable|
+        class.
+    latex_names : bool, optional keyword-only
+        Whether to format the parameters (column) names to latex -- e.g. 'H0' to
+        $$H_{0}$$.
+    **kwargs : Any
+        Passed to ``cls.write``.
+
+    Raises
+    ------
+    TypeError
+        If the optional keyword-argument 'cls' is not a subclass of |Table|.
+    ValueError
+        If the keyword argument 'format' is given and is not "ascii.html".
+
+    Notes
+    -----
+    A HTML file containing a Cosmology HTML table should have scripts enabling
+    MathJax.
+
+    ::
+        <script
+        src="https://polyfill.io/v3/polyfill.min.js?features=es6"></script>
+        <script type="text/javascript" id="MathJax-script" async
+            src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-chtml.js">
+        </script>
+    """
+    # Check that the format is 'ascii.html' (or not specified)
+    format = kwargs.pop("format", "ascii.html")
+    if format != "ascii.html":
+        raise ValueError(f"format must be 'ascii.html', not {format}")
+
+    # Set cosmology_in_meta as false for now since there is no metadata being kept
+    table = to_table(cosmology, cls=cls, cosmology_in_meta=False)
+
+    cosmo_cls = type(cosmology)
+    for name, col in table.columns.items():
+        param = getattr(cosmo_cls, name, None)
+        if not isinstance(param, Parameter) or param.unit in (None, u.one):
+            continue
+        # Replace column with unitless version
+        table.replace_column(name, (col << param.unit).value, copy=False)
+
+    # TODO! move the `latex_names` into `to_table`
+    if latex_names:
+        new_names = [_FORMAT_TABLE.get(k, k) for k in cosmology.__parameters__]
+        table.rename_columns(cosmology.__parameters__, new_names)
+
+    # Write HTML, using table I/O
+    table.write(file, overwrite=overwrite, format="ascii.html", **kwargs)
+
+
+def html_identify(origin, filepath, fileobj, *args, **kwargs):
+    """Identify if an object uses the HTML Table format.
+
+    Parameters
+    ----------
+    origin : Any
+        Not used.
+    filepath : str or Any
+        From where to read the Cosmology.
+    fileobj : Any
+        Not used.
+    *args : Any
+        Not used.
+    **kwargs : Any
+        Not used.
+
+    Returns
+    -------
+    bool
+        If the filepath is a string ending with '.html'.
+    """
+    return isinstance(filepath, str) and filepath.endswith(".html")
+
+
+# ===================================================================
+# Register
+
+readwrite_registry.register_reader("ascii.html", Cosmology, read_html_table)
+readwrite_registry.register_writer("ascii.html", Cosmology, write_html_table)
+readwrite_registry.register_identifier("ascii.html", Cosmology, html_identify)

--- a/astropy/cosmology/io/tests/test_.py
+++ b/astropy/cosmology/io/tests/test_.py
@@ -12,7 +12,7 @@ def test_expected_readwrite_io():
     """Test that ONLY the expected I/O is registered."""
 
     got = {k for k, _ in readwrite_registry._readers.keys()}
-    expected = {"ascii.ecsv"}
+    expected = {"ascii.ecsv", "ascii.html"}
 
     assert got == expected
 

--- a/astropy/cosmology/io/tests/test_html.py
+++ b/astropy/cosmology/io/tests/test_html.py
@@ -1,0 +1,256 @@
+# Licensed under a 3-clause BSD style license - see LICENSE.rst
+
+# THIRD PARTY
+import pytest
+
+import astropy.units as u
+from astropy.cosmology.io.html import _FORMAT_TABLE, read_html_table, write_html_table
+from astropy.cosmology.parameter import Parameter
+from astropy.table import QTable, Table, vstack
+from astropy.units.decorators import NoneType
+from astropy.utils.compat.optional_deps import HAS_BS4
+
+from .base import ReadWriteDirectTestBase, ReadWriteTestMixinBase
+
+###############################################################################
+
+
+class ReadWriteHTMLTestMixin(ReadWriteTestMixinBase):
+    """
+    Tests for a Cosmology[Read/Write] with ``format="ascii.html"``.
+    This class will not be directly called by :mod:`pytest` since its name does
+    not begin with ``Test``. To activate the contained tests this class must
+    be inherited in a subclass. Subclasses must dfine a :func:`pytest.fixture`
+    ``cosmo`` that returns/yields an instance of a |Cosmology|.
+    See ``TestCosmology`` for an example.
+    """
+
+    @pytest.mark.skipif(not HAS_BS4, reason="requires beautifulsoup4")
+    def test_to_html_table_bad_index(self, read, write, tmp_path):
+        """Test if argument ``index`` is incorrect"""
+        fp = tmp_path / "test_to_html_table_bad_index.html"
+
+        write(fp, format="ascii.html")
+
+        # single-row table and has a non-0/None index
+        with pytest.raises(IndexError, match="index 2 out of range"):
+            read(fp, index=2, format="ascii.html")
+
+        # string index where doesn't match
+        with pytest.raises(KeyError, match="No matches found for key"):
+            read(fp, index="row 0", format="ascii.html")
+
+    # -----------------------
+
+    @pytest.mark.skipif(not HAS_BS4, reason="requires beautifulsoup4")
+    def test_to_html_table_failed_cls(self, write, tmp_path):
+        """Test failed table type."""
+        fp = tmp_path / "test_to_html_table_failed_cls.html"
+
+        with pytest.raises(TypeError, match="'cls' must be"):
+            write(fp, format='ascii.html', cls=list)
+
+    @pytest.mark.parametrize("tbl_cls", [QTable, Table])
+    @pytest.mark.skipif(not HAS_BS4, reason="requires beautifulsoup4")
+    def test_to_html_table_cls(self, write, tbl_cls, tmp_path):
+        fp = tmp_path / "test_to_html_table_cls.html"
+        write(fp, format='ascii.html', cls=tbl_cls)
+
+    # -----------------------
+
+    @pytest.mark.skipif(not HAS_BS4, reason="requires beautifulsoup4")
+    def test_readwrite_html_table_instance(self, cosmo_cls, cosmo, read, write, tmp_path, add_cu):
+        """Test cosmology -> ascii.html -> cosmology."""
+        fp = tmp_path / "test_readwrite_html_table_instance.html"
+
+        # ------------
+        # To Table
+
+        write(fp, format="ascii.html")
+
+        # some checks on the saved file
+        tbl = QTable.read(fp)
+        # assert tbl.meta["cosmology"] == cosmo_cls.__qualname__  # metadata read not implemented
+        assert tbl["name"] == cosmo.name
+
+        # ------------
+        # From Table
+
+        tbl["mismatching"] = "will error"
+        tbl.write(fp, format="ascii.html", overwrite=True)
+
+        # tests are different if the last argument is a **kwarg
+        if tuple(cosmo._init_signature.parameters.values())[-1].kind == 4:
+            got = read(fp, format="ascii.html")
+
+            assert got.__class__ is cosmo_cls
+            assert got.name == cosmo.name
+            # assert "mismatching" not in got.meta # metadata read not implemented
+
+            return  # don't continue testing
+
+        # read with mismatching parameters errors
+        with pytest.raises(TypeError, match="there are unused parameters"):
+            read(fp, format="ascii.html")
+
+        # unless mismatched are moved to meta
+        got = read(fp, format="ascii.html", move_to_meta=True)
+        assert got == cosmo
+        # assert got.meta["mismatching"] == "will error" # metadata read not implemented
+
+        # it won't error if everything matches up
+        tbl.remove_column("mismatching")
+        tbl.write(fp, format="ascii.html", overwrite=True)
+        got = read(fp, format="ascii.html")
+        assert got == cosmo
+
+        # and it will also work if the cosmology is a class
+        # Note this is not the default output of ``write``.
+        # tbl.meta["cosmology"] = _COSMOLOGY_CLASSES[tbl.meta["cosmology"]] #
+        # metadata read not implemented
+        got = read(fp, format="ascii.html")
+        assert got == cosmo
+
+        got = read(fp)
+        assert got == cosmo
+
+    @pytest.mark.skipif(not HAS_BS4, reason="requires beautifulsoup4")
+    def test_rename_html_table_columns(self, read, write, tmp_path):
+        """Tests renaming columns"""
+        fp = tmp_path / "test_rename_html_table_columns.html"
+
+        write(fp, format="ascii.html", latex_names=True)
+
+        tbl = QTable.read(fp)
+
+        # asserts each column name has not been reverted yet
+        # For now, Cosmology class and name are stored in first 2 slots
+        for column_name in tbl.colnames[2:]:
+            assert column_name in _FORMAT_TABLE.values()
+
+        cosmo = read(fp, format="ascii.html")
+        converted_tbl = cosmo.to_format("astropy.table")
+
+        # asserts each column name has been reverted
+        # cosmology name is still stored in first slot
+        for column_name in converted_tbl.colnames[1:]:
+            assert column_name in _FORMAT_TABLE.keys()
+
+    @pytest.mark.skipif(not HAS_BS4, reason="requires beautifulsoup4")
+    @pytest.mark.parametrize("latex_names", [True, False])
+    def test_readwrite_html_subclass_partial_info(self, cosmo_cls, cosmo, read,
+                                                  write, latex_names, tmp_path, add_cu):
+        """
+        Test writing from an instance and reading from that class.
+        This works with missing information.
+        """
+        fp = tmp_path / "test_read_html_subclass_partial_info.html"
+
+        # test write
+        write(fp, format="ascii.html", latex_names=latex_names)
+
+        # partial information
+        tbl = QTable.read(fp)
+
+        # tbl.meta.pop("cosmology", None) # metadata not implemented
+        cname = "$$T_{0}$$" if latex_names else "Tcmb0"
+        del tbl[cname]  # format is not converted to original units
+        tbl.write(fp, overwrite=True)
+
+        # read with the same class that wrote fills in the missing info with
+        # the default value
+        got = cosmo_cls.read(fp, format="ascii.html")
+        got2 = read(fp, format="ascii.html", cosmology=cosmo_cls)
+        got3 = read(fp, format="ascii.html", cosmology=cosmo_cls.__qualname__)
+
+        assert (got == got2) and (got2 == got3)  # internal consistency
+
+        # not equal, because Tcmb0 is changed, which also changes m_nu
+        assert got != cosmo
+        assert got.Tcmb0 == cosmo_cls._init_signature.parameters["Tcmb0"].default
+        assert got.clone(name=cosmo.name, Tcmb0=cosmo.Tcmb0, m_nu=cosmo.m_nu) == cosmo
+        # but the metadata is the same
+        # assert got.meta == cosmo.meta # metadata read not implemented
+
+    @pytest.mark.skipif(not HAS_BS4, reason="requires beautifulsoup4")
+    def test_readwrite_html_mutlirow(self, cosmo, read, write, tmp_path, add_cu):
+        """Test if table has multiple rows."""
+        fp = tmp_path / "test_readwrite_html_mutlirow.html"
+
+        # Make
+        cosmo1 = cosmo.clone(name="row 0")
+        cosmo2 = cosmo.clone(name="row 2")
+        table = vstack([c.to_format("astropy.table") for c in (cosmo1, cosmo, cosmo2)],
+                       metadata_conflicts='silent')
+
+        cosmo_cls = type(cosmo)
+        if cosmo_cls == NoneType:
+            assert False
+
+        for n, col in zip(table.colnames, table.itercols()):
+            if n == "cosmology":
+                continue
+            param = getattr(cosmo_cls, n)
+            if not isinstance(param, Parameter) or param.unit in (None, u.one):
+                continue
+            # Replace column with unitless version
+            table.replace_column(n, (col << param.unit).value, copy=False)
+
+        table.write(fp, format="ascii.html")
+
+        # ------------
+        # From Table
+
+        # it will error on a multi-row table
+        with pytest.raises(ValueError, match="need to select a specific row"):
+            read(fp, format="ascii.html")
+
+        # unless the index argument is provided
+        got = cosmo_cls.read(fp, index=1, format="ascii.html")
+        # got = read(fp, index=1, format="ascii.html")
+        assert got == cosmo
+
+        # the index can be a string
+        got = cosmo_cls.read(fp, index=cosmo.name, format="ascii.html")
+        assert got == cosmo
+
+        # it's better if the table already has an index
+        # this will be identical to the previous ``got``
+        table.add_index("name")
+        got2 = cosmo_cls.read(fp, index=cosmo.name, format="ascii.html")
+        assert got2 == cosmo
+
+
+class TestReadWriteHTML(ReadWriteDirectTestBase, ReadWriteHTMLTestMixin):
+    """
+    Directly test ``read/write_html``.
+    These are not public API and are discouraged from use, in favor of
+    ``Cosmology.read/write(..., format="ascii.html")``, but should be
+    tested regardless b/c they are used internally.
+    """
+
+    def setup_class(self):
+        self.functions = {"read": read_html_table, "write": write_html_table}
+
+    @pytest.mark.skipif(not HAS_BS4, reason="requires beautifulsoup4")
+    def test_rename_direct_html_table_columns(self, read, write, tmp_path):
+        """Tests renaming columns"""
+
+        fp = tmp_path / "test_rename_html_table_columns.html"
+
+        write(fp, format="ascii.html", latex_names=True)
+
+        tbl = QTable.read(fp)
+
+        # asserts each column name has not been reverted yet
+        for column_name in tbl.colnames[2:]:
+            # for now, Cosmology as metadata and name is stored in first 2 slots
+            assert column_name in _FORMAT_TABLE.values()
+
+        cosmo = read(fp, format="ascii.html")
+        converted_tbl = cosmo.to_format("astropy.table")
+
+        # asserts each column name has been reverted
+        for column_name in converted_tbl.colnames[1:]:
+            # for now now, metadata is still stored in first slot
+            assert column_name in _FORMAT_TABLE.keys()

--- a/astropy/cosmology/tests/test_connect.py
+++ b/astropy/cosmology/tests/test_connect.py
@@ -9,8 +9,10 @@ from astropy import cosmology
 from astropy.cosmology import Cosmology, w0wzCDM
 from astropy.cosmology.connect import readwrite_registry
 from astropy.cosmology.io.tests import (
-    test_cosmology, test_ecsv, test_json, test_mapping, test_model, test_row, test_table, test_yaml)
+    test_cosmology, test_ecsv, test_html, test_json, test_mapping, test_model, test_row, test_table,
+    test_yaml)
 from astropy.table import QTable, Row
+from astropy.utils.compat.optional_deps import HAS_BS4
 
 ###############################################################################
 # SETUP
@@ -18,7 +20,13 @@ from astropy.table import QTable, Row
 cosmo_instances = cosmology.realizations.available
 
 # Collect the registered read/write formats.
-readwrite_formats = {"ascii.ecsv", "json"}
+#   (format, supports_metadata, has_all_required_dependencies)
+readwrite_formats = {
+    ("ascii.ecsv", True, True),
+    ("ascii.html", False, HAS_BS4),
+    ("json", True, True)
+}
+
 
 # Collect all the registered to/from formats. Unfortunately this is NOT
 # automatic since the output format class is not stored on the registry.
@@ -27,10 +35,14 @@ tofrom_formats = [("mapping", dict), ("yaml", str),
                   ("astropy.cosmology", Cosmology),
                   ("astropy.row", Row), ("astropy.table", QTable)]
 
+
 ###############################################################################
 
 
-class ReadWriteTestMixin(test_ecsv.ReadWriteECSVTestMixin, test_json.ReadWriteJSONTestMixin):
+class ReadWriteTestMixin(
+        test_ecsv.ReadWriteECSVTestMixin,
+        test_html.ReadWriteHTMLTestMixin,
+        test_json.ReadWriteJSONTestMixin):
     """
     Tests for a CosmologyRead/Write on a |Cosmology|.
     This class will not be directly called by :mod:`pytest` since its name does
@@ -40,14 +52,17 @@ class ReadWriteTestMixin(test_ecsv.ReadWriteECSVTestMixin, test_json.ReadWriteJS
     See ``TestReadWriteCosmology`` or ``TestCosmology`` for examples.
     """
 
-    @pytest.mark.parametrize("format", readwrite_formats)
-    def test_readwrite_complete_info(self, cosmo, tmp_path, format):
+    @pytest.mark.parametrize("format, metaio, has_deps", readwrite_formats)
+    def test_readwrite_complete_info(self, cosmo, tmp_path, format, metaio, has_deps):
         """
         Test writing from an instance and reading from the base class.
         This requires full information.
         The round-tripped metadata can be in a different order, so the
         OrderedDict must be converted to a dict before testing equality.
         """
+        if not has_deps:
+            pytest.skip("missing a dependency")
+
         fname = str(tmp_path / f"{cosmo.name}.{format}")
         cosmo.write(fname, format=format)
 
@@ -63,31 +78,35 @@ class ReadWriteTestMixin(test_ecsv.ReadWriteECSVTestMixin, test_json.ReadWriteJS
         got = Cosmology.read(fname, format=format)
 
         assert got == cosmo
-        assert dict(got.meta) == dict(cosmo.meta)
+        assert (not metaio) ^ (dict(got.meta) == dict(cosmo.meta))
 
-    @pytest.mark.parametrize("format", readwrite_formats)
-    def test_readwrite_from_subclass_complete_info(self, cosmo_cls, cosmo, tmp_path, format):
+    @pytest.mark.parametrize("format, metaio, has_deps", readwrite_formats)
+    def test_readwrite_from_subclass_complete_info(
+            self, cosmo_cls, cosmo, tmp_path, format, metaio, has_deps):
         """
         Test writing from an instance and reading from that class, when there's
         full information saved.
         """
+        if not has_deps:
+            pytest.skip("missing a dependency")
+
         fname = str(tmp_path / f"{cosmo.name}.{format}")
         cosmo.write(fname, format=format)
 
         # read with the same class that wrote.
         got = cosmo_cls.read(fname, format=format)
         assert got == cosmo
-        assert got.meta == cosmo.meta
+        assert (not metaio) ^ (dict(got.meta) == dict(cosmo.meta))
 
         # this should be equivalent to
         got = Cosmology.read(fname, format=format, cosmology=cosmo_cls)
         assert got == cosmo
-        assert got.meta == cosmo.meta
+        assert (not metaio) ^ (dict(got.meta) == dict(cosmo.meta))
 
         # and also
         got = Cosmology.read(fname, format=format, cosmology=cosmo_cls.__qualname__)
         assert got == cosmo
-        assert got.meta == cosmo.meta
+        assert (not metaio) ^ (dict(got.meta) == dict(cosmo.meta))
 
 
 class TestCosmologyReadWrite(ReadWriteTestMixin):
@@ -103,8 +122,11 @@ class TestCosmologyReadWrite(ReadWriteTestMixin):
 
     # ==============================================================
 
-    @pytest.mark.parametrize("format", readwrite_formats)
-    def test_write_methods_have_explicit_kwarg_overwrite(self, format):
+    @pytest.mark.parametrize("format, _, has_deps", readwrite_formats)
+    def test_write_methods_have_explicit_kwarg_overwrite(self, format, _, has_deps):
+        if not has_deps:
+            pytest.skip("missing a dependency")
+
         writer = readwrite_registry.get_writer(format, Cosmology)
         # test in signature
         sig = inspect.signature(writer)
@@ -113,11 +135,13 @@ class TestCosmologyReadWrite(ReadWriteTestMixin):
         # also in docstring
         assert "overwrite : bool" in writer.__doc__
 
-    @pytest.mark.parametrize("format", readwrite_formats)
-    def test_readwrite_reader_class_mismatch(self, cosmo, tmp_path, format):
+    @pytest.mark.parametrize("format, _, has_deps", readwrite_formats)
+    def test_readwrite_reader_class_mismatch(self, cosmo, tmp_path, format, _, has_deps):
         """Test when the reader class doesn't match the file."""
+        if not has_deps:
+            pytest.skip("missing a dependency")
 
-        fname = str(tmp_path / f"{cosmo.name}.{format}")
+        fname = tmp_path / f"{cosmo.name}.{format}"
         cosmo.write(fname, format=format)
 
         # class mismatch

--- a/docs/changes/cosmology/13075.feature.rst
+++ b/docs/changes/cosmology/13075.feature.rst
@@ -1,0 +1,2 @@
+Cosmology instance can be parsed from or converted to a HTML table using
+the new HTML methods in Cosmology's ``to/from_format`` I/O.

--- a/docs/cosmology/io.rst
+++ b/docs/cosmology/io.rst
@@ -33,9 +33,9 @@ Getting Started
 The |Cosmology| class includes two methods, |Cosmology.read| and
 |Cosmology.write|, that make it possible to read from and write to files.
 
-Currently the only registered ``read`` / ``write`` format is "ascii.ecsv",
-like for Table. Also, custom ``read`` / ``write`` formats may be registered
-into the Astropy Cosmology I/O framework.
+The registered ``read`` / ``write`` formats include "ascii.ecsv" and
+"ascii.html", like for Table. Also, custom ``read`` / ``write`` formats may be
+registered into the Astropy Cosmology I/O framework.
 
 Writing a cosmology instance requires only the file location and optionally,
 if the file format cannot be inferred, a keyword argument "format". Additional
@@ -66,6 +66,7 @@ To see a list of the available read/write file formats:
       Format   Read Write Auto-identify
     ---------- ---- ----- -------------
     ascii.ecsv  Yes   Yes           Yes
+    ascii.html  Yes   Yes           Yes
 
 This list will include both built-in and registered 3rd-party formats.
 

--- a/docs/whatsnew/5.2.rst
+++ b/docs/whatsnew/5.2.rst
@@ -51,6 +51,23 @@ cosmologies with their non-flat equivalents.
     True
 
 
+A cosmology can be parsed from or converted to a HTML table using
+the new HTML methods in Cosmology's ``to/from_format`` I/O.
+
+    >>> from astropy.cosmology import Planck18
+    >>> Planck18.write("planck18.html")
+
+The columns can be latex/mathjax formatted using the flag ``latex_names=True``;
+then if the following is added to the file's header, the column names will
+render nicely.::
+
+    <script
+    src="https://polyfill.io/v3/polyfill.min.js?features=es6"></script>
+    <script type="text/javascript" id="MathJax-script" async
+        src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-chtml.js">
+    </script>
+
+
 .. _whatsnew-5.2-coordinates:
 
 Topocentric ITRS Frame

